qcodes/tests/test_visa.py

```diff
@@ -109,6 +112,33 @@ def _make_mock_visa():
         mv.close()
 
 
+def test_visa_gc_closes_connection(caplog) -> None:
+    def use_magnet():
+        _ = AMIModel430(
+            "x",
+            address="GPIB::1::INSTR",
+            pyvisa_sim_file="AMI430.yaml",
+            terminator="\n",
+        )
+        assert list(Instrument._all_instruments.keys()) == ["x"]
+
+    # ensure that any unused instruments that have not been gced are gced before running
+    gc.collect()
+    caplog.clear()
+    with caplog.at_level(logging.INFO, logger="qcodes.instrument.visa"):
+        use_magnet()
+        gc.collect()
+    # at this stage the instrument created in use_magnet has gone out of scope
+    # and we have triggered an explicit gc so the weakref.finalize function has been triggered
+    # We test this
+    # and the instrument should no longer be in the instrument registry
+    assert list(Instrument._all_instruments.keys()) == []
```

Comment:
should we have an additional assert here on pyvisa.resources.?? (ResourceManager?) to ensure that there are no open handles?

Reply:
Why not, that seems like a good idea

Reply:
@astafan8  I could not find a way to get from the visa resource to the resource handler. I have implemented this by exposing the resource handler on the instrument which I think is probably a good idea anyway.

Reply:
sounds good!